src/common/StackStringStream.h

```diff
@@ -0,0 +1,165 @@
+ * License version 2.1, as published by the Free Software
+ * Foundation.	See file COPYING.
+ *
+ */
+
+#ifndef COMMON_STACKSTRINGSTREAM_H
+#define COMMON_STACKSTRINGSTREAM_H
+
+#include <boost/container/small_vector.hpp>
+
+#include <algorithm>
+#include <cstring>
+#include <iostream>
+#include <ostream>
+#include <sstream>
+#include <string_view>
+#include <vector>
+
+template<std::size_t SIZE>
+class StackStringBuf : public std::basic_streambuf<char>
+{
+public:
+  StackStringBuf() = default;
+  ~StackStringBuf() override = default;
+  StackStringBuf(const StackStringBuf&) = delete;
+  StackStringBuf& operator=(const StackStringBuf&) = delete;
+  StackStringBuf(StackStringBuf&& o)
+  : basic_streambuf<char>(std::move(o)),
+    vec(std::move(o.vec))
+  {
+    o.clear();
+  }
+
+  StackStringBuf& operator=(StackStringBuf&& o)
+  {
+    basic_streambuf<char>::operator=(std::move(o));
+    vec = std::move(o.vec);
+    o.clear();
+    return *this;
+  }
+
+  void push(std::string_view sv)
+  {
+    vec.reserve(vec.size() + sv.size());
+    vec.insert(vec.end(), sv.begin(), sv.end());
+  }
+
+  void clear()
+  {
+    vec.clear();
+  }
+
+  std::string_view strv() const
+  {
+    return std::string_view(vec.data(), vec.size());
+  }
+
+protected:
+  std::streamsize xsputn(const char *s, std::streamsize n)
+  {
+    push(std::string_view(s, n));
+    return n;
+  }
+
+  int overflow(int c)
+  {
+    if (traits_type::not_eof(c)) {
+      char str = traits_type::to_char_type(c);
+      push(std::string_view(&str, 1));
+      return c;
+    }
+    return EOF;
+  }
+
+private:
+
+  boost::container::small_vector<char, SIZE> vec;
+};
+
+class StackStringStream : public std::ostream
+{
+public:
+  StackStringStream() : std::ostream(&ssb) {}
+  StackStringStream(const StackStringStream& o) = delete;
+  StackStringStream& operator=(const StackStringStream& o) = delete;
+  StackStringStream(StackStringStream&& o)
+  : std::ostream(std::move(o)), ssb(std::move(o.ssb))
+  {}
+  StackStringStream& operator=(StackStringStream&& o)
+  {
+    std::ostream::operator=(std::move(o));
+    ssb = std::move(o.ssb);
+    return *this;
+  }
+  ~StackStringStream() override = default;
+
+  void clear() {
+    std::ostream::clear();
+    ssb.clear();
+  }
+
+  std::string_view strv() const {
+    return ssb.strv();
+  }
+
+private:
+  StackStringBuf<4096> ssb;
+};
+
+/* In an ideal world, we could use StackStringStream indiscriminately, but alas
+ * it's very expensive to construct/destruct. So, we cache them in a
+ * thread_local vector. DO NOT share these with other threads. The copy/move
+ * constructors are deliberately restrictive to make this more difficult to
+ * accidentally do.
+ */
+class CachedStackStringStream {
+public:
+  using osptr = std::unique_ptr<StackStringStream>;
+  using cache = std::vector<osptr>;
+
+  CachedStackStringStream() {
+    auto& c = get_cache();
+    if (c.empty()) {
+      c.emplace_back(std::make_unique<StackStringStream>());
```

Comment:
i worry a bit about using thread_local with std::vectors, because they'll grow to a peak capacity and never return space when the size goes back down. given a pathological workload, you could end up with lots of threads holding on to a lot of memory

if we already have to allocate the nodes themselves, can we use an intrusive list for this cache instead? that could eliminate the allocations from std::vector too

Reply:
Another option is to just have a limit on the size of the vector. Since we're talking about pathological cases, I'd rather just add a simple check.

I'll look at using an intrusive list instead.